RedHatInsights/rhsm-subscriptions
# SWATCH-2478: Add billable usage UUIDs to aggregates so we can update retry_on

Jira issue: SWATCH-2478

## Description

Our method for looking up billable_usage_remittances for retry_on updates was returning more than one result because the query was incorrect. We now have added a UUID to the records which can be used to find unique records instead so this PR is to update the dead letter logic to lookup based on UUID.

## Testing

IQE Test MR: 

### Setup

1. `podman-compose up`
2. mock the subscription service:  

`podman run -it --rm -p 8101:8080 --name wiremock -v $PWD/stub:/home/wiremock:z wiremock/wiremock:2.32.0 --verbose`

```
cat > stub/__files/swatch-2099.json <<EOF
{
    "errors" : [  { "code": "SUBSCRIPTIONS1006" }  ]
}
EOF
```

```
curl -X POST --data '{ "priority": 1, "request": { "urlPath": "/api/rhsm-subscriptions/v1/internal/subscriptions/azureUsageContext", "method": "GET"}, "response": { "status": 404, "bodyFileName": "swatch-2099.json", "headers":{"Content-Type":"application/json"} }}' http://localhost:8101/__admin/mappings/new
```
3. Insert existing remittances:
```
INSERT INTO public.billable_usage_remittance VALUES (NULL, 'org123', 'rhel-for-x86-els-payg', 'vCPUs', '2024-04', 'Premium', 'Production', 'azure', 'test;test', 1, '2024-04-23 12:01:57.670879+00', NULL, '5487b805-944f-45ec-b3c9-27f427a6f155', 'AZURE', '80e74ff7-6453-4f4b-976a-1a1b60b9055f', NULL, NULL, NULL);
INSERT INTO public.billable_usage_remittance VALUES (NULL, 'org123', 'rhel-for-x86-els-payg', 'vCPUs', '2024-04', 'Premium', 'Production', 'azure', 'test;test', 1, '2024-04-23 11:01:57.670879+00', NULL, '5487b805-944f-45ec-b3c9-27f427a6f155', 'AZURE', '051eb959-a65a-4b67-9402-7ab2af94b82c', NULL, NULL, NULL);
```
4. Start azure producer:
`SERVER_PORT=8002 QUARKUS_MANAGEMENT_ENABLED=false SWATCH_INTERNAL_SUBSCRIPTION_ENDPOINT=http://localhost:8101 ./gradlew :swatch-producer-azure:quarkusDev`
5. Start billable-usage service:
`SERVER_PORT=8001 KSTREAM_BILLABLE_USAGE_AGGREGATION_WINDOW_DURATION=1s KSTREAM_BILLABLE_USAGE_AGGREGATION_GRACE_DURATION=0s QUARKUS_MANAGEMENT_ENABLED=false ./gradlew :swatch-billable-usage:quarkusDev
`
Test Steps
1. Run following curl command twice, waiting at least one second between each:
```
curl 'http://localhost:8401/api/swatch-producer-azure/internal/azure/billable_usage'   -H 'accept: */*'   -H 'Content-Type: application/json'   -d '{
    "org_id":"org123",
    "id":"5487b805-944f-45ec-b3c9-27f427a6f155",
    "billing_provider":"azure",
    "billing_account_id":"test;test",
    "snapshot_date":"2024-04-23T11:01:57.670879Z",
    "product_id":"rhel-for-x86-els-payg",
    "sla":"Premium",
    "usage":"Production",
    "metric_id":"vCPUs",
    "value":1.0,
    "uuid": "051eb959-a65a-4b67-9402-7ab2af94b82c"
  }'
```
1. Start tally-service with: `DEV_MODE=true ./gradlew :bootRun`

Verification:
1. Check that retry_after is populated on record:
`select retry_after from public.billable_usage_remittance where uuid='051eb959-a65a-4b67-9402-7ab2af94b82c';`
